README.md
# VELOBS 08
Application mobile pour le site web VELOBS.

VELOBS est donc un site web qui permet aux cyclistes de signaler les problèmes et de proposer des améliorations sur le réseau cyclable. Il est développé depuis 2009 par "L’association toulousaine 2 pieds 2 roues".

Depuis 2016, les sources de VELOBS sont en libre accès sur Github : https://github.com/2p2r/velobs_web.



## INSTALLATION
Download zip project and setup it up with cordova.



## TROUBLESHOOTING

-



## LICENSE

Copyright © 2015-2019 by Sebastien TRIONFINI <prof@trionfini.com>. All rights reserved.

CHANGELOG.md
## v2.0

Nouveautés

- Stockage des informations en local : adresse mail + serveur

- Réduction du nombre de pages de 6 à 3

- Bar de statut toujours visible : pour voir le réseau disponible

Améliorations

- Interface graphique plus pratique : menu de navigation plus propre, bouton plus gros et dynamique

- Réinitialisation des champs après l'envoi de l'observation et retour au menu

- Message de retour sur la confirmation de l'envoi plus précis

Bugs

- Adresse email moins restrictive

- Boîte d'alerte native : suppression du index.html



### Description

Enregistrez les problèmes rencontrés par les cyclistes dans les Ardennes.

Placez le curseur sur la carte où se trouve l'observation que vous voulez soumettre.

Vous pouvez ajouter un commentaire et/ou une photo, ou bien créez une nouvelle observation en spécifiant le type de problème, sa localisation, sa description votre proposition pour résoudre ce problème et ajoutez une photo si besoin (recommandé pour faciliter le traitement de votre observation) avant envoi.

Les bénévoles de Ma Ville à Vélo 08, du Codep08 modèrent (vérifient) vos observations, mises à disposition des services techniques des responsables des voiries.



## v1.0

- First commit



### Promotional Text

Première version disponible sur iOS !

N'hésitez pas à nous remonter les problèmes que vous pourriez rencontrer.

Merci d'avance.



### Description

Enregistrez les problèmes rencontrés par les cyclistes dans les Ardennes, ou mettez des observations à jour.

Placez le curseur sur la carte où se trouve l'observation que vous voulez soumettre. Vérifiez qu'elle n'existe pas déjà.

Le cas échéant, vous pouvez ajouter un commentaire et/ou une photo, ou bien créez une nouvelle observation en spécifiant le type de problème, sa localisation, sa description votre proposition pour résoudre ce problème et ajoutez une photo si besoin (recommandé pour faciliter le traitement de votre observation) avant envoi.

Les bénévoles de Ma Ville à Vélo 08, du Codep08 modèrent (vérifient) vos observations, mises à disposition des services techniques des responsables des voiries



### Keywords

velobs, velo, bicyclette, cycle



### Support URL

https://www.2p2r.org/articles-divers/page-sommaire/article/velobs?lang=fr



### Marketing URL

-



### Copyright 

2019 Sebastien TRIONFINI



### Contributing

Contributor, add your name under as following :

Sebastien TRIONFINI <prof@trionfini.com>



www/sw.js
const staticCacheName = "cache-v1"
const assets = [
  '/',
  '/index.html'
]

self.addEventListener('install', (e) => {
    e.waitUntil(
      caches.open(staticCacheName).then((cache) => {
        cache.addAll(assets)
      })
    )
});

self.addEventListener('fetch', (event) => {
  event.respondWith(
    caches.match(event.request)
      .then(function(response) {
        // Cache hit - return response
        if (response) {
          return response;
        }

        // IMPORTANT: Cloner la requête.
        // Une requete est un flux et est à consommation unique
        // Il est donc nécessaire de copier la requete pour pouvoir l'utiliser et la servir
        var fetchRequest = event.request.clone();

        return fetch(fetchRequest).then(
          function(response) {
            if (!response || response.status !== 200 || response.type !== 'basic') {
              return response;
            }

            // IMPORTANT: Même constat qu'au dessus, mais pour la mettre en cache
            var responseToCache = response.clone();

            caches.open(staticCacheName)
              .then(function(cache) {
                cache.put(event.request, responseToCache);
              });

            return response;
          }
        );
      })
  );
});

self.addEventListener('activate', (e) => {
  e.waitUntil(
    caches.keys().then((keys) => {
      return Promise.add(
        keys.filter((key) => key != staticCacheName).map((key) => caches.delete(key))
      );
    })
  );
});


www/index.js
var app = {
    // Error with loading map multiple time
    mapIsLoaded: false,

    // Application Constructor
    initialize: function() {
        document.addEventListener('deviceready', this.onDeviceReady.bind(this), false);
    },
    bindEvents: function() {
        // Here we register our callbacks for the lifecycle events we care about
        document.addEventListener('deviceready', this.onDeviceReady, false);
        document.addEventListener('pause', this.onPause, false);
        document.addEventListener('resume', this.onResume, false);
    },

    // deviceready Event Handler
    //
    // Bind any cordova events here. Common events are:
    // 'pause', 'resume', etc.
    onDeviceReady: function() {
        // Provides access to the device's audio, image, and video capture capabilities
        //console.log(navigator.device.capture);

        // Run the first page
        this.firstPage();
    },
    firstPage: function(){
        var configIsSetup = window.localStorage.getItem("LSsetup");

        if(configIsSetup){
            configUpdate(configIsSetup);

            $('#newOBS').show();
            this.router("_home");
        } else {
            $('#newOBS').hide();
            this.router("_config");
        }
    },
    router: function(page){
        if(page === "_1map"){
            $('.slide').hide();
            $('.slide'+page).show();

            if(this.mapIsLoaded){
                $( "div.infos" ).fadeIn( 300 ).delay( 2500 ).fadeOut( 400 );
            } else {
                navigator.geolocation.getCurrentPosition(onSuccess, onError);
                $( "div.infos" ).fadeIn( 300 ).delay( 2500 ).fadeOut( 400 );

                this.mapIsLoaded = true;
            }
        } else {
            $('.slide').hide();
            $('.slide'+page).show();
        }
    }
};
 
    if ("serviceWorker" in navigator) {
        navigator.serviceWorker
        .register("sw.js")
        .then(() => console.log("sw registered"))
        .catch((err) => console.log(err))
    }

app.initialize();
app.dispatchEvent("deviceready");

www/main.js
navigator.geolocation.getCurrentPosition(function(position) {
    faireQqc(position.coords.latitude, position.coords.longitude);
  });

  var watchID = navigator.geolocation.watchPosition(function(position) {
    faireQqc(position.coords.latitude, position.coords.longitude);
  });

  navigator.geolocation.clearWatch(watchID);

  function geo_success(position) {
    do_something(position.coords.latitude, position.coords.longitude);
  }
  
  function geo_error() {
    alert("Sorry, no position available.");
  }
  
  var geo_options = {
    enableHighAccuracy: true,
    maximumAge        : 30000,
    timeout           : 27000
  };
  
  var wpid = navigator.geolocation.watchPosition(geo_success, geo_error, geo_options);

  function errorCallback(error) {
    alert('ERROR(' + error.code + '): ' + error.message);
  };
  
  function prompt(window, pref, message, callback) {
    let branch = Components.classes["@mozilla.org/preferences-service;1"]
                           .getService(Components.interfaces.nsIPrefBranch);

    if (branch.getPrefType(pref) === branch.PREF_STRING) {
        switch (branch.getCharPref(pref)) {
        case "always":
            return callback(true);
        case "never":
            return callback(false);
        }
    }

    let done = false;

    function remember(value, result) {
        return function() {
            done = true;
            branch.setCharPref(pref, value);
            callback(result);
        }
    }

    let self = window.PopupNotifications.show(
        window.gBrowser.selectedBrowser,
        "geolocation",
        message,
        "geo-notification-icon",
        {
            label: "Share Location",
            accessKey: "S",
            callback: function(notification) {
                done = true;
                callback(true);
            }
        }, [
            {
                label: "Always Share",
                accessKey: "A",
                callback: remember("always", true)
            },
            {
                label: "Never Share",
                accessKey: "N",
                callback: remember("never", false)
            }
        ], {
            eventCallback: function(event) {
                if (event === "dismissed") {
                    if (!done) callback(false);
                    done = true;
                    window.PopupNotifications.remove(self);
                }
            },
            persistWhileVisible: true
        });
}

prompt(window,
       "extensions.foo-addon.allowGeolocation",
       "Foo Add-on wants to know your location.",
       function callback(allowed) { alert(allowed); });


function configUpdate(haveLocalStorageSetup) {
    if(haveLocalStorageSetup){
        document.getElementById('input_mail').value = window.localStorage.getItem("LSmail");

        var chosenServer = window.localStorage.getItem("LSserver");

        if(chosenServer === "https://velobs08.ouvaton.org/"){
            document.getElementById('prodServer').checked = true;
        } else if (chosenServer === "https://egnom.pro/VELOBS08/"){
            document.getElementById('devServer').checked = true;
        } else {
            document.getElementById('customServer').value = chosenServer;
        }

        getSubCategory(chosenServer);
    } else {
        var inputMail = $('#input_mail').val();

        //email test
        if (/\S+@\S+\.\S+/.test(inputMail)) {
            //URL test
            if (document.getElementById('prodServer').checked) {
                configURL = document.getElementById('prodServer').value;
    
                window.localStorage.setItem("LSmail", inputMail);
                window.localStorage.setItem("LSserver", configURL);
                window.localStorage.setItem("LSsetup", true);
                getSubCategory(configURL);
                $('#newOBS').show();
                app.router("_home");
                
            } else if (document.getElementById('devServer').checked) {
                configURL = document.getElementById('devServer').value;
    
                window.localStorage.setItem("LSmail", inputMail);
                window.localStorage.setItem("LSserver", configURL);
                window.localStorage.setItem("LSsetup", true);
                getSubCategory(configURL);
                $('#newOBS').show();
                app.router("_home");
    
            } else {
                if(document.getElementById('customServer').value !== ""){
                    configURL = document.getElementById('customServer').value;
    
                    window.localStorage.setItem("LSmail", inputMail);
                    window.localStorage.setItem("LSserver", configURL);
                    window.localStorage.setItem("LSsetup", true);
                    getSubCategory(configURL);
                    $('#newOBS').show();
                    app.router("_home");
                } else {
                    //alert("URL non valide");
                    navigator.notification.alert("URL non valide");
                }
            }
        } else {
            $('#newOBS').hide();
            //alert("Adresse email non valide");
            navigator.notification.alert("Adresse email non valide");
        }
    }

}

// Sous catégories
function getSubCategory(URL) {
    var xhr = new XMLHttpRequest();

    xhr.onload = function(){
        if (xhr.status >= 200 && xhr.status < 300) {
            //console.log('success!', xhr);

            //Delete previous subcategory may already added
            document.getElementById('obs').innerHTML = "";

            var sub = xhr.response;
            sub = sub.substring(1,sub.length);
            sub = sub.substring(0,sub.length - 1);
            sub = JSON.parse(sub);

            var i = sub["subcategory"].length;
            for (var a = 0; a < i;a++){
                document.getElementById('obs').innerHTML += '<option value="' + sub["subcategory"][a]["id_subcategory"] + '">' + sub["subcategory"][a]["lib_subcategory"] + '</option>' ;
            };
        } else {
            //console.log('The request failed!');
            //alert("La demande a échoué");
            navigator.notification.alert("La demande a échoué");
            
            
        };
        //console.log('This always runs...');
    };
    xhr.open('GET', URL + 'lib/php/public/getSubCategory.php');
    xhr.send();
};

//  Photo to show from selected file
function readURL(input) {
	if (input.files && input.files[0]) {
			var reader = new FileReader();

			reader.onload = function (e) {
					$('#showSelectedImage')
                            .attr('src', e.target.result)
                            .show()
                    ;
			};

			reader.readAsDataURL(input.files[0]);
	}
}

// Form to generate and to send 
function sendForm(){
    var formElement = new FormData();

    var chosenServer = window.localStorage.getItem("LSserver");

    // Waiting from the server but not in the APP
    formElement.append("task", "CREATEPUBLICPOI");
    //formElement.append("num_poi", document.getElementsByName("num_poi")[0].value);
    //formElement.append("tel_poi", document.getElementsByName("tel_poi")[0].value);
    //formElement.append("adherent_poi", document.getElementsByName("adherent_poi")[0].value);

    formElement.append("mail_poi", $('#input_mail').val());
    formElement.append("rue_poi", $('#Rue').val());
    formElement.append("desc_poi", $('#Description').val());
    formElement.append("prop_poi", $('#Proposition').val());
    formElement.append("latitude_poi", lat);
    formElement.append("longitude_poi", lng);
    formElement.append("subcategory_id_subcategory", $('#obs').val());

    // Server require $_FILE['photo-path'] even if empty, why ?
    if(document.getElementById("photoToSend").files && document.getElementById("photoToSend").files[0]){
        formElement.append("photo-path", document.getElementById("photoToSend").files[0]);
    } else {
        formElement.append("photo-path", new File([], ""));
    }

    var xhr = new XMLHttpRequest();
    xhr.onload = function () {
        if (xhr.status >= 200 && xhr.status < 300) {
            //console.log('success!', xhr);
            var responseJSON = JSON.parse(xhr.responseText);
            //console.log(responseJSON);

            if(responseJSON.success === true){
                //alert(responseJSON.ok);
                navigator.notification.alert(responseJSON.ok);

                //Reset form
                location.reload();
            } else if (responseJSON.success === false){
                //alert(responseJSON.pb);
                navigator.notification.alert(responseJSON.pb);
            }
        } else {
            //console.log('The request failed!');
            //alert("La demande a échoué");
            navigator.notification.alert("La demande a échoué");
        }
        //console.log('This always runs...');
    };
    xhr.open("POST", chosenServer + "lib/php/admin/database.php");
    xhr.send(formElement);
}